Plutarch/Api/V1.hs

```diff
@@ -0,0 +1,572 @@
+
+instance PIsDataRepr PScriptPurpose where
+  type
+    PIsDataReprRepr PScriptPurpose =
+      '[ '[PCurrencySymbol]
+       , '[PTxOutRef]
+       , '[PStakingCredential]
+       , '[PDCert]
+       ]
+
+  pmatchRepr dat f =
+    pmatchDataRepr dat $
+      DRHCons (f . PMinting) $
+        DRHCons (f . PSpending) $
+          DRHCons (f . PRewarding) $
+            DRHCons
+              (f . PCertifying)
+              DRHNil
+
+---------- Scripts
+
+newtype PDatum (s :: S) = PDatum (Term s PData)
+
+instance PUnsafeLiftDecl PDatum where type PLifted PDatum = Plutus.Datum
+deriving via (DerivePConstantViaNewtype Plutus.Datum PDatum PData) instance (PConstant Plutus.Datum)
+
+newtype PRedeemer (s :: S) = PRedeemer (Term s PData)
+
+instance PUnsafeLiftDecl PRedeemer where type PLifted PRedeemer = Plutus.Redeemer
+deriving via (DerivePConstantViaNewtype Plutus.Redeemer PRedeemer PData) instance (PConstant Plutus.Redeemer)
+
+newtype PDatumHash (s :: S)
+  = PDatumHash (Term s PByteString)
+
+instance PUnsafeLiftDecl PDatumHash where type PLifted PDatumHash = Plutus.DatumHash
+deriving via (DerivePConstantViaNewtype Plutus.DatumHash PDatumHash PByteString) instance (PConstant Plutus.DatumHash)
+
+newtype PStakeValidatorHash (s :: S)
+  = PStakeValidatorHash (Term s PByteString)
+
+instance PUnsafeLiftDecl PStakeValidatorHash where type PLifted PStakeValidatorHash = Plutus.StakeValidatorHash
+deriving via (DerivePConstantViaNewtype Plutus.StakeValidatorHash PStakeValidatorHash PByteString) instance (PConstant Plutus.StakeValidatorHash)
+
+newtype PRedeemerHash (s :: S)
+  = PRedeemerHash (Term s PByteString)
+
+instance PUnsafeLiftDecl PRedeemerHash where type PLifted PRedeemerHash = Plutus.RedeemerHash
+deriving via (DerivePConstantViaNewtype Plutus.RedeemerHash PRedeemerHash PByteString) instance (PConstant Plutus.RedeemerHash)
+
+newtype PValidatorHash (s :: S)
+  = PValidatorHash (Term s PByteString)
+
+instance PUnsafeLiftDecl PValidatorHash where type PLifted PValidatorHash = Plutus.ValidatorHash
+deriving via (DerivePConstantViaNewtype Plutus.ValidatorHash PValidatorHash PByteString) instance (PConstant Plutus.ValidatorHash)
+
+---------- Value
+
+newtype PTokenName (s :: S)
+  = PTokenName (Term s PByteString)
+  deriving newtype (Semigroup, Monoid)
+instance PUnsafeLiftDecl PTokenName where type PLifted PTokenName = Plutus.TokenName
+deriving via
+  (DerivePConstantViaNewtype Plutus.TokenName PTokenName PByteString)
+  instance
+    (PConstant Plutus.TokenName)
+
+newtype PCurrencySymbol (s :: S)
+  = PCurrencySymbol (Term s PByteString)
+instance PUnsafeLiftDecl PCurrencySymbol where type PLifted PCurrencySymbol = Plutus.CurrencySymbol
+deriving via
+  (DerivePConstantViaNewtype Plutus.CurrencySymbol PCurrencySymbol PByteString)
+  instance
+    (PConstant Plutus.CurrencySymbol)
+
+newtype PValue (s :: S)
+  = PValue (Term s (PMap PCurrencySymbol (PMap PTokenName PInteger)))
```

Comment:
Another important newtype lacking instances...

Reply:
I believe @TotallyNotChase is working on this FWIW

Reply:
I haven't started generalizing what I already have yet. Will get to it later today unless someone else picks it up.

Reply:
This has yet to be tackled, though #151 partially solves the lacking instances problem.

Reply:
`PValue` now has both `PlutusType` and `PIsData` instances through #161 

It's currently missing `PConstant`/`PLift` instances (need those for `PMap` too). I assume this might also be resolved by #159